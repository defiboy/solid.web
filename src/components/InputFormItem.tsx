import React from 'react'

import { Form, Input } from 'antd'
import { FieldProps } from 'formik'

const FormItem = Form.Item

export interface InputProps {
  label: string
  placeHolder?: string
}

type InputOwnProps = FieldProps & InputProps

export const InputFormItem: React.FC<InputOwnProps> = ({ label, field, form, placeHolder }: InputOwnProps) => { // TODO: Type this and make it generic
  return (
    <FormItem
      htmlFor={field.name}
      label={label}
      hasFeedback={!!form.errors.name}
      validateStatus={form.errors[field.name] && 'error'}
      help={form.errors[field.name]}>
      <Input id={field.name} value={field.value} size="large" type="text" {...field} data-testid={`input-${field.name}`} placeholder={placeHolder} />
    </FormItem>
  )
}
