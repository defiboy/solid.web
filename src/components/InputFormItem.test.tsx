import React from 'react';
import { Formik, FormikErrors, Field } from 'formik';
import { render, fireEvent } from '@testing-library/react'

import { InputFormItem } from './InputFormItem'

describe('InputFormItem', () => {

    const onSubmitMockHandler = jest.fn()

    type GenericType = {
        name: string,
        url: string
    }

    const initialValues: GenericType = {
        name: '',
        url: ''
    }

    const renderWithWrapperForm = (element: React.ReactElement) => {
        return render(<Formik
            initialValues={initialValues}
            onSubmit={onSubmitMockHandler}
            validate={(items: GenericType) => {
                const errors: FormikErrors<GenericType> = {}
                if (!items.name) {
                    errors.name = 'Required'
                }
                if (!items.url) {
                    errors.url = 'Required'
                }
                return errors
            }}>
            {element}
        </Formik>)
    }

    it('renders an input', async () => {
        const { getByLabelText, getByTestId } = renderWithWrapperForm(
            <Field
                name="url"
                render={(innerProps: any) => (
                    <InputFormItem {...innerProps} label="Blockchain URL" placeHolder="JSON RPC endpoint" />
                )}
            />
        )

        expect(getByLabelText(/Blockchain URL/i)).toBeInTheDocument()
        expect(getByTestId('input-url')).toBeInTheDocument()
    })

    it('renders an input with label and placeholder', () => {
        const placeHolder = "JSON RPC endpoint"
        const label = "Blockchain URL"
        const name = "url"

        const { getByTestId } = renderWithWrapperForm(
            <Field
                name={name}
                render={(innerProps: any) => (
                    <InputFormItem {...innerProps} label={label} placeHolder={placeHolder} />
                )}
            />
        )

        const element = getByTestId('input-url')

        expect(element).toHaveAttribute('placeHolder', placeHolder)
        expect(element).toHaveAttribute('type', 'text')
        expect(element).toHaveAttribute('name', name)
        expect(element).toHaveAttribute('id', name)
        expect(element).toHaveAttribute('value', '')
        expect(element).toHaveAttribute('class', 'ant-input ant-input-lg')
    })

    it('can be set a new value to the input', () => {
        const placeHolder = "JSON RPC endpoint"
        const label = "Blockchain URL"
        const name = "url"

        const { getByTestId } = renderWithWrapperForm(
            <Field
                name={name}
                render={(innerProps: any) => (
                    <InputFormItem {...innerProps} label={label} placeHolder={placeHolder} />
                )}
            />
        )

        const element = getByTestId('input-url')

        expect(element).toHaveAttribute('value', '')

        const newValue = "newValue"

        fireEvent.change(element, {
            target: {
                value: newValue
            }
        })

        expect(element).toHaveAttribute('value', newValue)
    })

    it('should match snapshot', () => {
        const placeHolder = "JSON RPC endpoint"
        const label = "Blockchain URL"
        const name = "url"

        const { container } = renderWithWrapperForm(
            <Field
                name={name}
                render={(innerProps: any) => (
                    <InputFormItem {...innerProps} label={label} placeHolder={placeHolder} />
                )}
            />
        )

        expect(container).toMatchSnapshot()
    })

})
