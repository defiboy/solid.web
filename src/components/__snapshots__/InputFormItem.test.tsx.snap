// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputFormItem should match snapshot 1`] = `
<div>
  <div
    class="ant-row ant-form-item"
  >
    <div
      class="ant-col ant-form-item-label"
    >
      <label
        class=""
        for="url"
        title="Blockchain URL"
      >
        Blockchain URL
      </label>
    </div>
    <div
      class="ant-col ant-form-item-control-wrapper"
    >
      <div
        class="ant-form-item-control"
      >
        <span
          class="ant-form-item-children"
        >
          <input
            class="ant-input ant-input-lg"
            data-testid="input-url"
            id="url"
            name="url"
            placeholder="JSON RPC endpoint"
            type="text"
            value=""
          />
        </span>
      </div>
    </div>
  </div>
</div>
`;
